### PR TITLE
[Datasets] Add UDF passthrough args to map_batches().

### PR DESCRIPTION
**Pull request by clarkzinzow**
Users often have UDFs that require arguments only known at Dataset creation time, and sometimes these arguments may be large objects better suited for shared zero-copy access via the object store. One example is batch inference on a large on-CPU model using the actor pool strategy: without putting the large model into the object store and sharing across actors on the same node, each actor worker will need its own copy of the model.

Unfortunately, we can't rely on closing over these object refs, since the object ref will then be serialized in the exported function/class definition, causing the object to be indefinitely pinned and therefore leaked. It's much cleaner to instead link these object refs in as actor creation and task arguments. This PR adds support for threading such object refs through as actor creation and task arguments and supplying the concrete values to the UDFs.

## TODOs

- [x] Add docs/example to "Transforming Datasets" user guide

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/_internal/compute.py [hunk 1]
```diff
@@ -38,7 +38,17 @@ def _apply(
         block_list: BlockList,
         clear_input_blocks: bool,
         name: Optional[str] = None,
+        fn_args: Optional[List[Any]] = None,
+        fn_kwargs: Optional[Dict[str, Any]] = None,
+        fn_constructor_args: Optional[List[Any]] = None,
+        fn_constructor_kwargs: Optional[Dict[str, Any]] = None,
     ) -> BlockList:
+        assert fn_constructor_args is None and fn_constructor_kwargs is None
```
Comment (ericl) [inline python/ray/data/_internal/compute.py:46]: ?

**Review by ericl — COMMENTED** [review]
Code LGTM. Could we also update the docs (especially any uses of callable classes)?

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -257,6 +262,10 @@ def map_batches(
         batch_size: Optional[int] = 4096,
         compute: Union[str, ComputeStrategy] = None,
         batch_format: str = "native",
+        fn_args: Optional[List[Any]] = None,
```
Comment (jianoaix) [inline python/ray/data/dataset.py:265]: This is half of the parameter list, how does it compare to wrapping them in a FnParams class? Not sure if it's a good idea in Python, but it's good practice in C++ (e.g. https://abseil.io/tips/173).

Comment (clarkzinzow) [reply python/ray/data/dataset.py:265]: Yeah threading these individual arg collections is annoying, unfortunately there are a few reasons that keeping these as independent arguments is required:

1. When passing these to remote functions (actor or plain tasks), these need to be top-level arguments in order for object refs to be properly resolved to their concrete values, which requires destructuring them before passing them into the functions.
2. When passing these to user UDFs, we need to destructure them in order to cleanly support the default case of a UDF that only takes a `batch` argument: by destructuring these passthrough arguments when calling the UDF, if not are specified, it will be equivalent to a `udf(batch)` call; if you passed them directly as e.g. `fn_args` or a `FnParams` object, if the UDF isn't expecting that argument, then this will fail.
3. When passing these to the block transformation function in the compute strategy, we need to have it result in no arguments passed to block transformation functions that don't support arg passthroughs (e.g. all APIs other than `.map_batches()`). This wouldn't work if we were passing in an explicit `FnParams` object.

Basically this boils down to wanting Datasets operations and UDFs being able to opt-in to receiving these passthrough args, and the passthrough args needing to be top-level arguments in the (actor) tasks.

Comment (jianoaix) [reply python/ray/data/dataset.py:265]: Sg, thanks for the detailed explanation!

File: python/ray/data/dataset.py [hunk 3]
```diff
@@ -228,12 +235,22 @@ def map(
             ray_remote_args: Additional resource requirements to request from
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
+        if isinstance(fn, CallableClass) and (
+            compute is None
+            or compute == "tasks"
+            or isinstance(compute, TaskPoolStrategy)
+        ):
+            raise ValueError(
```
Comment (jianoaix) [inline python/ray/data/dataset.py:243]: If this a must, why not just overwrite whatever from input args? Maybe just warn that it's overwritten in that case.

File: python/ray/data/dataset.py [hunk 4]
```diff
@@ -322,10 +364,42 @@ def map_batches(
...
+                "``compute`` must be specified when using a CallableClass, and must "
+                f"specify the actor compute strategy, but got: {compute}"
+                'For example, use ``compute="actors"`` or '
+                "``compute=ActorPoolStrategy(min, max)``."
+            )
+
+        if fn_constructor_args is not None or fn_constructor_kwargs is not None:
```
Comment (jianoaix) [inline python/ray/data/dataset.py:379]: I like these sanity checks, can we add some unit tests for them?

File: python/ray/data/_internal/plan.py [hunk 5]
```diff
@@ -399,14 +408,24 @@ class OneToOneStage(Stage):
...
         self.block_fn = block_fn
         self.compute = compute or "tasks"
         self.ray_remote_args = ray_remote_args or {}
+        self.fn = fn
+        self.fn_args = fn_args
+        self.fn_kwargs = fn_kwargs
+        self.fn_constructor_args = fn_constructor_args
+        self.fn_constructor_kwargs = fn_constructor_kwargs
 
     def can_fuse(self, prev: Stage):
```
Comment (jianoaix) [inline python/ray/data/_internal/plan.py:430]: Can you add some tests for fusion case? The current tests have just one mapping operation, add some more to invoke this code.

File: python/ray/data/tests/test_dataset.py [hunk 6]
```diff
@@ -1873,6 +1916,62 @@ def test_map_batch(ray_start_regular_shared, tmp_path):
...
+        assert a == 1
+        assert b == 2
+        return b * batch + a
+
+    ds = ray.data.read_parquet(str(tmp_path))
+    ds2 = ds.map_batches(
+        udf,
+        batch_size=1,
+        batch_format="pandas",
+        fn_args=(ray.put(1),),
```
Comment (jianoaix) [inline python/ray/data/tests/test_dataset.py:1934]: Add cases for passing args via just one of them?

Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:1934]: Hmm doesn't the args propagating successfully from both imply that the args will successfully propagate for each individually? I can't see how a failure scenario could arise where passing both succeeds but passing e.g. only `fn_args` would fail.

File: python/ray/data/dataset.py (hunk 3 above)
Comment (jianoaix) [reply python/ray/data/dataset.py:245]: The default is None, and this is raising exception on that default, that's weird...
The clearer semantics to me is we don't make this check, i.e. allow CallableClass to run with task pool, but log a warning on potential setup cost. This way the API is sane in terms of functionality, the difference only in performance.

Comment (clarkzinzow) [reply python/ray/data/dataset.py:245]: That sounds good to me, but I don't want to add that change in semantics as a driveby change in this PR, so let's surface this change in the default internally and open a follow-up PR.

Comment (jianoaix) [reply python/ray/data/dataset.py:245]: The issue is exactly that this part of code is  changing the API semantics:
- Before: map(CallableClass, TaskPoolStrategy) works
- After: it will raise exception

Comment (clarkzinzow) [reply python/ray/data/dataset.py:245]: No, as I said above, `map(CallableClass, TaskPoolStrategy)` current fails in master.

>Also to be clear, raising an error if a callable class and the default task pool strategy is provided is the current behavior, I just moved the error from within the map tasks to the API level for better error UX.

Comment (clarkzinzow) [reply python/ray/data/dataset.py:245]: >This worked on a previous release, so it must have changed lately.

We removed support for that since stateful tasks are an anti-pattern that hit weird reference counting edge cases and resulted in long-lived workers hogging clusters: https://github.com/ray-project/ray/pull/23708

> Then the action item here is to revise the documentation to reflect this new semantics.

The master API docs call out that callable classes are only supported for the actor compute model: https://docs.ray.io/en/master/data/package-ref.html#ray.data.Dataset.map_batches

And I add a dedicated [callable classes section](https://ray--24870.org.readthedocs.build/en/24870/data/transforming-datasets.html#callable-class-udfs) to the "Transforming Datasets" feature guide in the NumPy batch PR: https://github.com/ray-project/ray/pull/24870

So I think that docs are covered.

Comment (jianoaix) [reply python/ray/data/dataset.py:245]: Thanks for having them documented. For the docstring here, since this constraint is on "compute", it looks more natural to document it at "compute" arg than "fn" arg.

File: python/ray/data/tests/test_dataset.py (hunk 6 above)
Comment (jianoaix) [reply python/ray/data/tests/test_dataset.py:1981]: Maybe, the issue is we never know and things can happen in unexpected ways. If we take an agnostic view about internals and just look at the new map_batches() API, and try to design test cases for it, I'd think this would be an angle to cover.

**Comment by ericl** [comment]
LGTM at high level. Will leave detailed review to Jian here.

File: python/ray/data/dataset.py (hunk 3 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:245]: Hmm but isn't the `fn` arg the best place to document it? Users are going to be looking at this UDF-first, where they have a UDF that involves some state initialization (e.g. cached loading of a model), so they'll look at the `fn` arg to see if it takes a class. If they see that a `CallableClass` is accepted without a note about only supporting the actor compute strategy, then they might miss a note in the `compute` arg docs warning that `compute="actors"` must be used in order to use a `CallableClass` UDF.

How about we document it in both places?

File: python/ray/data/tests/test_dataset.py (hunk 6 above)
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:1981]: Tests are geared towards the semantics of the API, and the semantics are that these are independent passthrough variables, so it's hard to see how the implementation could deviate to the point where a single test asserting that both are passed through would pass while the individual cases would fail. There's a development and maintenance cost to each test written, so I was trying to be pragmatic here.

Anyway, I'll add individual arg test coverage to resolve this.

File: python/ray/data/dataset.py (hunk 3 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:250]: I ended up adding a note about the constraint to both args in the map docstrings.

**Comment by clarkzinzow** [comment]
@jianoaix Feedback has been implemented, this is ready for another pass!

File: python/ray/data/_internal/plan.py [hunk 7]
```diff
@@ -415,6 +435,23 @@ def can_fuse(self, prev: Stage):
         # not the other way around. The latter will be used as the compute if fused.
         if is_task_compute(self.compute) and prev.compute != self.compute:
             return False
+        if prev.fn_constructor_args != self.fn_constructor_args:
+            return False
+        if prev.fn_constructor_kwargs != self.fn_constructor_kwargs:
+            return False
```
Comment (jianoaix) [inline python/ray/data/_internal/plan.py:441]: What's the reason they have to be the same? If the prev is a task and self is CallableClass/actor, or if they are both CallableClasses but run different fn (or same fn with different construction args), these seem fine?

File: python/ray/data/dataset.py [hunk 8]
```diff
@@ -248,17 +270,27 @@ def transform(block: Block) -> Iterable[Block]:
...
 
     def map_batches(
         self,
-        fn: Union[CallableClass, Callable[[BatchType], BatchType]],
+        fn: BatchUDF,
         *,
         batch_size: Optional[int] = 4096,
         compute: Union[str, ComputeStrategy] = None,
         batch_format: str = "native",
+        fn_args: Optional[List[Any]] = None,
```
Comment (jianoaix) [inline python/ray/data/dataset.py:290]: Does this need to be a Tuple per Python's convention?

**Review by jianoaix — COMMENTED** [review]
Mostly LG

File: python/ray/data/_internal/plan.py (hunk 7 above)
Comment (clarkzinzow) [reply python/ray/data/_internal/plan.py:441]: For the multiple callable classes case (difference callable classes, or same callable class with different construction args), we can't support this since we can't initialize multiple callable classes in the actor constructor. Callable classes are initialized once, in the actor constructor when an actor worker is created, and we can only initialize one callable class at a time since we need to provide `fn_constructor_args` and `fn_constructor_kwargs` as flat arguments so object references get properly resolved in the actor creation task.

For the normal function + callable class case, this is actually supposed to be supported (`fuse()` is designed to support it), so I think that this fuse check is too strict. Let me fix that!

File: python/ray/data/dataset.py (hunk 8 above)
Comment (clarkzinzow) [reply python/ray/data/dataset.py:290]: Yep good catch!

Comment (clarkzinzow) [reply python/ray/data/dataset.py:290]: Ah actually this can be any iterable, but I thought that it'd be easiest for the user if it was a list.

File: python/ray/data/_internal/plan.py (hunk 7 above)
Comment (jianoaix) [reply python/ray/data/_internal/plan.py:441]: It'd be great to comment the first case for the line that performs the check, and add a test case for them.

File: python/ray/data/_internal/plan.py [hunk 9]
```diff
@@ -393,20 +405,96 @@ def has_computed_output(self) -> bool:
         )
 
 
+def _pack_args(
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/_internal/plan.py:408]: This is packaging it into a flat positional arg list so any Ray object futures are properly resolved when these arguments are passed into a Ray task. We still can't pass a `FnParams` data class around as a required argument since we want Datasets operations and UDFs to be able to opt-in to receiving these passthrough args, as mentioned here: https://github.com/ray-project/ray/pull/25613#discussion_r893840918

File: python/ray/data/tests/test_dataset.py [hunk 10]
```diff
@@ -1789,8 +1882,204 @@ def test_map_batch(ray_start_regular_shared, tmp_path):
...
+        fn_constructor_kwargs={"b": ray.put(2)},
+    )
+    assert ds2._dataset_format() == "pandas"
+    ds_list = ds2.take()
+    values = [s["one"] for s in ds_list]
+    assert values == [3, 5, 7]
+    values = [s["two"] for s in ds_list]
+    assert values == [5, 7, 9]
+
+    # Test callable chain.
```
[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:2021]: I'll separate these call chain tests into another test and look at adding a stage fusion test in `test_optimize.py`!

[the remark this answers is not in the document]
Comment (clarkzinzow) [reply python/ray/data/tests/test_dataset.py:2031]: Extra args tests split out into new test and stage fusion test added.

**Review by jianoaix — APPROVED** [review]

**Comment by clarkzinzow** [comment]
@xwjiang2010 @krfricke I haven't been able to reproduce this Tuner test failure locally, but it's failed in the CI twice in a row: https://buildkite.com/ray-project/ray-builders-pr/builds/36294#01818932-b7c0-4bca-b779-bc41c8a3cc72

 ```
           with self.assertRaises(TuneError):
  >           tuner.fit()
  E           AssertionError: TuneError not raised
```

The captured stdout shows a worker dying:
```
2022-06-22 03:09:53,962	WARNING worker.py:1726 -- A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffff463b06118c2b18b2871339f301000000 Worker ID: f6e8c2761e19169236f39196924f6d43af44e8a55a456c9543af40b9 Node ID: 9b38f22bc2411f4f7e5bf4a5e26cad2a456d1b87ccd72ba1597efae2 Worker IP address: 172.18.0.3 Worker port: 35125 Worker PID: 25514 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.
--
  | 2022-06-22 03:09:53,964	ERROR trial_runner.py:918 -- Trial XGBoostTrainer_c7a9a_00000: Error processing event.
  | NoneType: None
  | (XGBoostTrainer pid=25514) [2022-06-22 03:09:53,949 C 25514 25514] core_worker.cc:849:  Check failed: has_owner Object IDs generated randomly (ObjectID.from_random()) or out-of-band (ObjectID.from_binary(...)) cannot be serialized because Ray does not know which task will create them. If this was not how your object ID was generated, please file an issue at https://github.com/ray-project/ray/issues/: 1507e51846dad045ffffffffffffffffffffffff0100000002000000
  | (XGBoostTrainer pid=25514) *** StackTrace Information ***
  | (XGBoostTrainer pid=25514)     ray::SpdLogMessage::Flush()
  | (XGBoostTrainer pid=25514)     ray::RayLog::~RayLog()
  | (XGBoostTrainer pid=25514)     ray::core::CoreWorker::GetOwnershipInfo()
  | (XGBoostTrainer pid=25514)     __pyx_pw_3ray_7_raylet_10CoreWorker_89serialize_object_ref()
  | (XGBoostTrainer pid=25514)     _PyMethodDef_RawFastCallKeywords
  | (XGBoostTrainer pid=25514)
```

I can't think of anything in this PR that could affect this test other than [this object ref reducer](https://github.com/ray-project/ray/pull/25613/files#diff-5a812d24fc3efd599284bb40dcbd2c6627bbe324e1c89a6e623b4dccf8e5de44R3336-R3344) in the Datasets lineage serialization, but this shouldn't be hit here AFAICT.

Any ideas?

**Comment by clarkzinzow** [comment]
>also if I am not mistaken, the code exercised by the time the error is thrown is the same as the other test - test_tuner_with_xgboost_trainer. Is that one failing? Doesn't seem like so.

Yeah, that test is passing in CI, and the only difference between the two tests is the `FailureInjectionCallback`...

**Comment by clarkzinzow** [comment]
Interesting, I was able to reproduce when running the entire `test_tuner.py` test module instead of the individual test, going to debug locally now.

**Comment by clarkzinzow** [comment]
Hmm @xwjiang2010 it looks like removing the out-of-band object ref reducer, added to catch out-of-band serialization of object refs when serializing datasets lineage, is the culprit; removing this causes the tests to pass. https://github.com/ray-project/ray/pull/25613/files#diff-5a812d24fc3efd599284bb40dcbd2c6627bbe324e1c89a6e623b4dccf8e5de44R3336-R3344
